ENIGMAsystem/SHELL/Widget_Systems/Win32/dialogs.cpp

```diff
@@ -558,37 +558,51 @@ int get_color(int defcolor, bool advanced)
 
 string get_directory(string dname, string caption)
 {
-//NOTE: This uses the Windows Vista or later file chooser, which is different than the one used by GM8 and lower
-//because I could not find out which one it uses, since IFileDialog is used by both wxWidgets and QtFramework
-//and there doesn't appear to be a standard file picker for XP or lower in the Windows API except SHBrowseForFolder that is
-//used by Game Maker for get_directory_alt
-  IFileDialog* fileDialog;
-  CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&fileDialog));
+  //NOTE: This uses the Windows Vista or later file chooser, which is different than the one used by GM8 and lower
+  //because I could not find out which one it uses, since IFileDialog is used by both wxWidgets and QtFramework
+  //and there doesn't appear to be a standard file picker for XP or lower in the Windows API except SHBrowseForFolder that is
+  //used by Game Maker for get_directory_alt
+  IFileDialog *selectDirectory;
+  CoCreateInstance(CLSID_FileOpenDialog, NULL, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(&selectDirectory));
 
   DWORD options;
-  fileDialog->GetOptions(&options);
-  options &= ~FOS_FILEMUSTEXIST;
-  options &= ~FOS_PATHMUSTEXIST;
-  fileDialog->SetOptions(options | FOS_PICKFOLDERS);
-  //TODO: Set default directory to dname
-  //fileDialog->SetDefaultFolder(std::wstring(dname.begin(), dname.end()).c_str());
-  fileDialog->SetTitle(std::wstring(caption.begin(), caption.end()).c_str());
-
-  fileDialog->Show(enigma::hWnd);
-
-  string res = "";
-  IShellItem *psi;
-
-  if (SUCCEEDED(fileDialog->GetResult(&psi))) {
-    LPWSTR wideres;
-    psi->GetDisplayName(SIGDN_DESKTOPABSOLUTEPARSING, &wideres);
-    psi->Release();
-
-    std::wstring wstr = wideres;
-    res = string(wstr.begin(), wstr.end());
+  selectDirectory->GetOptions(&options);
+  selectDirectory->SetOptions(options | FOS_PICKFOLDERS | FOS_NOCHANGEDIR | FOS_FORCEFILESYSTEM);
+
+  tstring tstr_dname = widen(dname);
+  LPWSTR szFilePath = (wchar_t *)tstr_dname.c_str();
+
+  IShellItem *pItem = nullptr;
+  HRESULT hr = ::SHCreateItemFromParsingName(szFilePath, nullptr, IID_PPV_ARGS(&pItem));
+
+  if (SUCCEEDED(hr)) {
+    LPWSTR szName = nullptr;
+    hr = pItem->GetDisplayName(SIGDN_NORMALDISPLAY, &szName);
+    if (SUCCEEDED(hr)) {
+      selectDirectory->SetFolder(pItem);
+      ::CoTaskMemFree(szName);
+    }
+    pItem->Release();
+  }
+
+  if (caption.empty()) selectDirectory->SetTitle(L"Select Directory");
+  else selectDirectory->SetTitle(std::wstring(caption.begin(), caption.end()).c_str());
+  selectDirectory->Show(enigma::hWnd);
+
+  pItem = nullptr;
+  hr = selectDirectory->GetResult(&pItem);
+
+  if (SUCCEEDED(hr)) {
+    LPWSTR wstr_result;
+    pItem->GetDisplayName(SIGDN_DESKTOPABSOLUTEPARSING, &wstr_result);
+    pItem->Release();
+
+    string str_result;
+    str_result = string_replace_all(shorten(wstr_result) + "\\", "\\\\", "\\");
```

Comment:
Why are we doing this again? I forgot about why we were doing these escapes?

Reply:
It's to add the final slash at the end. This is new behavior in GMStudio for directory related stuff. If the slash is added to a drive root, there would be a double slash. The string_replace_All removes the duplicate in such case to return a prettier string.

Reply:
Where do we stand on that for the other functions? Are the other functions in Win32 doing that yet? If not, please don't do it for now. I can't remember because I thought it died in review. I'd rather that decision be made all at once to all of the dialogs, but in general I can't remember what the situation was with the Win32 widgets.

Reply:
The only function in Win32 widgets remaining that doesn't do this is get_directory_alt which I plan to also make that change to in a different pr because it just so happens that function you wrote is also broken (the root argument doesn't work).

Reply:
That being said, all of our other widget systems include the last slash so imo it would be best to make these write-once, compile-anywhere.

Reply:
Is `get_open_filename` doing this? I can't seem to see where it does it.

Reply:
Then please create a helper function for it now like you did in the other places: https://github.com/enigma-dev/enigma-dev/pull/1493/files#r245889217

That way if somebody wants they can easily change it in just one spot. Honestly the same helper function could have been used across systems here but that's not too much of a biggy.

Reply:
Actually, there does seem to be an existing helper in there for this, please just reuse it.

Reply:
I forgot about it, will do.

Reply:
Resolved: 52b8ae95e91073b650573928c1f6d11f8f2ea7ec

You fixed it.